cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,305 @@
+    :param Vmax: The value of the upper bound: the value of V for which U + U_offset == 0.
+    :return: expr with a singularity exception around sp, if sp is not None else return the original expr.
+    """
+    f_Vmin = expr.xreplace({V: Vmin})
+    f_Vmax = expr.xreplace({V: Vmax})
+    return Piecewise((f_Vmin + ((V - Vmin) / (Vmax - Vmin)) * (f_Vmax - f_Vmin),
+                      Abs(V - sp) < Abs((Vmax - Vmin) / 2)), (expr, True))
+
+
+def _float_dummies(expr):
+    """Turns floats back into Quantity dummies to be in line with the rest of the sympy equations"""
+    return expr.xreplace({f: Quantity(f, 'dimensionless') for f in expr.atoms(Float)})
+
+
+def _is_negative_power(expr):
+    try:
+        return isinstance(expr, Pow) and bool(expr.args[1] < 0)
+    except TypeError:  # if this is a power with variables still in it we can't determine if it's negative
+        pass
+    return False
+
+
+def _solve_real(u, V):
+    """Gives the values of V for which u == 0 """
+    u = optimize(u, (_POW_OPT, ))  # make sure powers of ints are represented as ints
+    result = solveset(u, V, domain=S.Reals)
+    # The resul could be an intersection with Reals, if a custom exp function is used. We assume the result is real.
+    if isinstance(result, Intersection) and result.args[0] == S.Reals:
+        result = result.args[1]
+    return result
+
+
+def _get_singularity(expr, V, U_offset, exp_function):
+    """
+    Finds singularities in equations of the form:
+       - ``U / (exp(U) - 1.0)``
+       - ``U / (1.0 - exp(U))``
+       - ``(exp(U) - 1.0) / U``
+       - ``(1.0 - exp(U)) / U``
+
+    In addition to the singularity itself, this functions returns a lower and upper bounds as follows:
+    - sp is the singularity point: the value of V for which U == 0
+    - Vmin is the value of V for which U - U_offset == 0
+    - Vmax is the value of V for which U + U_offset == 0:
+    :return: (Vmin, Vmax, sp)
+    """
+
+    # Create "wildcards", that act as catch-all (*) when matching expressions
+    # https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.match
+    P = Wild('P', real=True, exclude=[V])
+    Z = Wild('Z', real=True)
+    U_wildcard = Wild('U_wildcard', real=True, include=[V])
+    SP_wildcard = Wild('SP_wildcard', real=True)
+    singularities = []
+
+    def check_U_match(m, sp):
+        assert m is None or m[Z] != 0
+        return m is not None and \
+            (sp == m[SP_wildcard] or (isinstance(sp, Float) and
+             isinstance(m[SP_wildcard], Float) and isclose(m[SP_wildcard], sp)))
+
+    # find all fractions and sperate numerator and denominator
+    # the denominator is all args where a **-1
+    numerator = [a for a in expr.args if not _is_negative_power(a)]
+    denominator = [Pow(a.args[0], - a.args[1]) for a in expr.args if _is_negative_power(a)]
+
+    # Not a division or does not have exp
+    if len(denominator) == 0 or len(numerator) == 0 or not expr.has(exp_function):
```

Comment:
```suggestion
    if len(denominator) == 0 or len(numerator) == 0:
```

Reply:
Instead of this check, you could just create the numerator and denominator (like below) and then check that both are functions of V?
Feels like that would be cheaper than starting the pattern matching before checking that?

Reply:
I just ran some performance tests, when through all equations in 2 models and for the rhs tried the following 100 times and then averaged the timings:
x = eq.rhs.has(exp_function)
x = eq.rhs.atoms(exp_function)
x = V in eq.rhs.free_symbols
Average timings for the 2 models I tries:
5.717277526855469e-07 1.3425486428397042e-05 6.567750658307756e-06
1.73602785382952e-06 1.3979775565011159e-05 7.481302533830915e-06
So the .has is a pretty quick operation

Reply:
Having looked at the sorts of equations we get there's plenty which have V but no exp, but I've seen very few that do have exp not not V